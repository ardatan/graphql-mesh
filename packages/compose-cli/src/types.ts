import { DocumentNode, GraphQLSchema } from 'graphql';
import { Logger } from '@graphql-mesh/types';
import { fetch as defaultFetch } from '@whatwg-node/fetch';

export interface MeshComposeCLIConfig {
  /**
   * The output destination of the resulting composed GraphQL schema.
   * By default, the CLI will write the result to stdout.
   */
  output?: string;
  subgraphs: MeshComposeCLISubgraphConfig[];
  transforms?: MeshComposeCLITransformConfig[];
  additionalTypeDefs?: string | DocumentNode | (string | DocumentNode)[];
  fetch?: typeof defaultFetch;
  cwd?: string;
}

export interface MeshComposeCLISubgraphConfig {
  sourceHandler: MeshComposeCLISourceHandlerDef;
  transforms?: MeshComposeCLITransformConfig[];
}

export type MeshComposeCLISourceHandlerDef = (ctx: LoaderContext) => {
  name: string;
  schema$: Promise<GraphQLSchema> | GraphQLSchema;
};

export type MeshComposeCLITransformConfig = (input: GraphQLSchema, ...args: any[]) => GraphQLSchema;

export interface LoaderContext {
  fetch: typeof defaultFetch;
  cwd: string;
  logger: Logger;
}

/**
 * Type helper for defining the config.
 */
export function defineConfig(config: MeshComposeCLIConfig) {
  return config;
}
