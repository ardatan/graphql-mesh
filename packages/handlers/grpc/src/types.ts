import {
  GraphQLType,
  GraphQLObjectType,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLBoolean,
} from 'graphql';

import {
  BigIntResolver as GraphQLBigInt,
} from 'graphql-scalars';

export interface GrpcGraphqlSchemaConfiguration {
  endpoint: string;
  protoFilePath: string;
  serviceName: string;
  packageName: string;
}

export interface TypeMapping {
  [key: string]: GraphQLType;
}

// https://developers.google.com/protocol-buffers/docs/proto3#scalar
// https://www.apollographql.com/docs/apollo-server/schemas/types.html
export const GRPC_GQL_TYPE_MAPPING: TypeMapping = {
  int32: GraphQLInt,
  int64: GraphQLBigInt, // TODO: https://github.com/graphql/graphql-js/issues/292
  float: GraphQLFloat,
  double: GraphQLFloat,
  string: GraphQLString,
  bool: GraphQLBoolean,
};

export const typeDefinitionCache: TypeMapping = {
  ServerStatus: new GraphQLObjectType({
    name: 'ServerStatus',
    description: 'status of the server',
    fields: () => ({
      status: {
        type: GraphQLString,
        descripton: 'status string',
      },
    }),
  }),
};
