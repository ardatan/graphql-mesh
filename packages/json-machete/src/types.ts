export type { JSONSchema, JSONSchemaObject } from '@json-schema-tools/meta-schema';
