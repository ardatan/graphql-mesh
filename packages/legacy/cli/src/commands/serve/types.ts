import type { MeshHTTPHandler } from '@graphql-mesh/http';
import type { MeshInstance } from '@graphql-mesh/runtime';
import type { YamlConfig } from '@graphql-mesh/types';

export interface ServerStartOptions {
  meshHTTPHandler: MeshHTTPHandler;
  getBuiltMesh: () => Promise<MeshInstance>;
  sslCredentials: YamlConfig.ServeConfig['sslCredentials'];
  graphqlPath: string;
  hostname: string;
  port: number;
}

export interface ServerStartResult {
  stop: VoidFunction;
}
