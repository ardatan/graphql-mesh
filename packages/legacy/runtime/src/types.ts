import type { DocumentNode, ExecutionResult } from 'graphql';
import type {
  GraphQLOperation,
  KeyValueCache,
  Logger,
  MeshFetch,
  MeshHandler,
  MeshMerger,
  MeshPlugin,
  MeshPubSub,
  MeshTransform,
  YamlConfig,
} from '@graphql-mesh/types';
import type { TerminateHandler } from '@graphql-mesh/utils';
import type { IResolvers, MaybeAsyncIterable, MaybePromise, Source } from '@graphql-tools/utils';
import type { MESH_CONTEXT_SYMBOL } from './constants.js';
import type { MeshInstance } from './get-mesh.js';

export type GetMeshOptions = {
  sources: MeshResolvedSource[];
  transforms?: MeshTransform[];
  additionalTypeDefs?: DocumentNode[];
  additionalResolvers?: IResolvers | IResolvers[];
  cache: KeyValueCache;
  pubsub?: MeshPubSub;
  merger: MeshMerger;
  logger?: Logger;
  additionalEnvelopPlugins?: MeshPlugin<any>[];
  documents?: Source[];
  fetchFn?: MeshFetch;
};

export type MeshResolvedSource = {
  name: string;
  handler: MeshHandler;
  transforms?: MeshTransform[];
};

export type ExecuteMeshFn<TData = any, TVariables = any, TContext = any, TRootValue = any> = (
  document: GraphQLOperation<TData, TVariables>,
  variables?: TVariables,
  context?: TContext,
  rootValue?: TRootValue,
  operationName?: string,
) => MaybePromise<ExecutionResult<TData>>;

export type SubscribeMeshFn<TVariables = any, TContext = any, TRootValue = any, TData = any> = (
  document: GraphQLOperation<TData, TVariables>,
  variables?: TVariables,
  context?: TContext,
  rootValue?: TRootValue,
  operationName?: string,
) => MaybePromise<MaybeAsyncIterable<ExecutionResult<TData>>>;

export type MeshContext = {
  [MESH_CONTEXT_SYMBOL]: true;
} & { pubsub: MeshPubSub; cache: KeyValueCache; logger: Logger; fetch: MeshFetch };

export interface ServeMeshOptions {
  baseDir: string;
  getBuiltMesh: () => Promise<MeshInstance>;
  logger: Logger;
  rawServeConfig: YamlConfig.Config['serve'];
  argsPort?: number;
  playgroundTitle?: string;
  registerTerminateHandler(handler: TerminateHandler): void;
}

export type MeshExecutor = <TData, TVariables, TContext, TRootValue>(
  documentOrSDL: GraphQLOperation<TData, TVariables>,
  variables?: TVariables,
  context?: TContext,
  rootValue?: TRootValue,
  operationName?: string,
) => Promise<TData | AsyncIterable<TData>>;
