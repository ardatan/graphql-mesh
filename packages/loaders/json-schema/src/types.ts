import { OperationTypeNode } from 'graphql';
import { JSONSchema } from 'json-machete';

export interface JSONSchemaBaseOperationConfig {
  type: OperationTypeNode;
  field: string;
  description?: string;

  responseSchema?: string | JSONSchema;
  responseSample?: any;
  responseTypeName?: string;

  argTypeMap?: Record<string, string>;
}

export interface JSONSchemaBaseOperationConfigWithJSONRequest extends JSONSchemaBaseOperationConfig {
  requestSchema?: string | JSONSchema;
  requestSample?: any;
  requestTypeName?: string;
}

export type HTTPMethod = 'GET' | 'HEAD' | 'POST' | 'PUT' | 'DELETE' | 'CONNECT' | 'OPTIONS' | 'TRACE' | 'PATCH';

export interface JSONSchemaHTTPBaseOperationConfig extends JSONSchemaBaseOperationConfig {
  path: string;
  method?: HTTPMethod;

  headers?: Record<string, string>;
}

export interface JSONSchemaHTTPJSONOperationConfig
  extends JSONSchemaHTTPBaseOperationConfig,
    JSONSchemaBaseOperationConfigWithJSONRequest {}

export interface JSONSchemaPubSubOperationConfig extends JSONSchemaBaseOperationConfigWithJSONRequest {
  pubsubTopic: string;
}

export interface JSONSchemaHTTPBinaryConfig extends JSONSchemaHTTPBaseOperationConfig {
  path: string;
  method?: HTTPMethod;
  requestTypeName?: string;
  binary: true;
}

export type JSONSchemaOperationConfig =
  | JSONSchemaHTTPJSONOperationConfig
  | JSONSchemaHTTPBinaryConfig
  | JSONSchemaPubSubOperationConfig;
