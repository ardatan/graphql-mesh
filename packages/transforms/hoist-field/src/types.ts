import { YamlConfig } from '@graphql-mesh/types';

export type HoistFieldTransformFieldPathConfig = YamlConfig.HoistFieldTransformFieldConfigObject['pathConfig'][0];
