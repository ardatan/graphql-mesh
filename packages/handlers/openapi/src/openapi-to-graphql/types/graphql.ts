// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: openapi-to-graphql
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * Custom type definitions for GraphQL.
 */

import {
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLInputObjectType,
  GraphQLList,
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLResolveInfo,
} from 'graphql';

export enum GraphQLOperationType {
  Query,
  Mutation,
  // TODO: Subscription
}

export type GraphQLType =
  | GraphQLObjectType
  | GraphQLInputObjectType
  | GraphQLList<any>
  | GraphQLUnionType
  | GraphQLEnumType
  | GraphQLScalarType;

type Arg = {
  type: any;
  description?: string;
};

export type Args = {
  [key: string]: Arg;
};

export type ResolveFunction = (root: any, args: any, ctx: any, info: GraphQLResolveInfo) => Promise<any> | any;

export type Field = {
  type: GraphQLType;
  resolve?: ResolveFunction;
  args?: Args;
  description: string;
};
