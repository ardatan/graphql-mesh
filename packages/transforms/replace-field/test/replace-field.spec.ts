import { join } from 'path';
import { execute, parse, printSchema, GraphQLObjectType } from 'graphql';
import InMemoryLRUCache from '@graphql-mesh/cache-inmemory-lru';
import { MeshPubSub } from '@graphql-mesh/types';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { PubSub } from 'graphql-subscriptions';

import ReplaceFieldTransform from '../src';

describe('replace-field', () => {
  const schemaDefs = /* GraphQL */ `
    type Query {
      books: BooksApiResponse
    }

    type BooksApiResponse {
      books: [Book]
    }

    type Book {
      title: String!
      author: Author!
    }

    type Author {
      name: String!
      age: Int!
    }
  `;
  let cache: InMemoryLRUCache;
  let pubsub: MeshPubSub;
  const baseDir: string = undefined;

  beforeEach(() => {
    cache = new InMemoryLRUCache();
    pubsub = new PubSub();
  });

  it('should replace correctly field with resolver function', async () => {
    const transform = new ReplaceFieldTransform({
      config: {
        replacements: [
          {
            from: {
              type: 'Query',
              field: 'books',
            },
            to: {
              type: 'BooksApiResponse',
              field: 'books',
            },
            scope: 'hoistValue',
          },
        ],
      },
      cache,
      pubsub,
      baseDir,
      apiName: '',
      syncImportFn: require,
    });
    const schema = makeExecutableSchema({
      typeDefs: schemaDefs,
      resolvers: {
        Query: {
          books: () => ({
            books: [{ title: 'abc' }, { title: 'def' }],
          }),
        },
      },
    });
    const transformedSchema = transform.transformSchema(schema);

    expect(transformedSchema.getType('BooksApiResponse')).toBeUndefined();
    expect((transformedSchema.getType('Query') as GraphQLObjectType).getFields().books.type.toString()).toBe('[Book]');
    expect(printSchema(transformedSchema)).toMatchSnapshot();

    const result = await execute({
      schema: transformedSchema,
      document: parse(/* GraphQL */ `
        {
          books {
            title
          }
        }
      `),
    });
    expect(result.data.books).toEqual([{ title: 'abc' }, { title: 'def' }]);
  });

  it('should replace correctly field without resolver function', async () => {
    const transform = new ReplaceFieldTransform({
      config: {
        replacements: [
          {
            from: {
              type: 'Book',
              field: 'author',
            },
            to: {
              type: 'Author',
              field: 'name',
            },
            scope: 'hoistValue',
          },
        ],
      },
      cache,
      pubsub,
      baseDir,
      apiName: '',
      syncImportFn: require,
    });
    const schema = makeExecutableSchema({
      typeDefs: schemaDefs,
      resolvers: {
        Query: {
          books: () => ({
            books: [
              { title: 'abc', author: { name: 'abra' } },
              { title: 'def', author: { name: 'cadabra' } },
            ],
          }),
        },
      },
    });
    const transformedSchema = transform.transformSchema(schema);

    expect(transformedSchema.getType('Author')).toBeUndefined();
    expect((transformedSchema.getType('Book') as GraphQLObjectType).getFields().author.type.toString()).toBe('String!');
    expect(printSchema(transformedSchema)).toMatchSnapshot();

    const result = await execute({
      schema: transformedSchema,
      document: parse(/* GraphQL */ `
        {
          books {
            books {
              title
              author
            }
          }
        }
      `),
    });
    expect(result.data.books.books).toEqual([
      { title: 'abc', author: 'abra' },
      { title: 'def', author: 'cadabra' },
    ]);
  });

  it('should replace correctly mtultiple fields with and without resolver function', async () => {
    const transform = new ReplaceFieldTransform({
      config: {
        replacements: [
          {
            from: {
              type: 'Query',
              field: 'books',
            },
            to: {
              type: 'BooksApiResponse',
              field: 'books',
            },
            scope: 'hoistValue',
          },
          {
            from: {
              type: 'Book',
              field: 'author',
            },
            to: {
              type: 'Author',
              field: 'name',
            },
            scope: 'hoistValue',
          },
        ],
      },
      cache,
      pubsub,
      baseDir,
      apiName: '',
      syncImportFn: require,
    });
    const schema = makeExecutableSchema({
      typeDefs: schemaDefs,
      resolvers: {
        Query: {
          books: () => ({
            books: [
              { title: 'abc', author: { name: 'abra' } },
              { title: 'def', author: { name: 'cadabra' } },
            ],
          }),
        },
      },
    });
    const transformedSchema = transform.transformSchema(schema);

    expect(transformedSchema.getType('BooksApiResponse')).toBeUndefined();
    expect(transformedSchema.getType('Author')).toBeUndefined();
    expect((transformedSchema.getType('Query') as GraphQLObjectType).getFields().books.type.toString()).toBe('[Book]');
    expect((transformedSchema.getType('Book') as GraphQLObjectType).getFields().author.type.toString()).toBe('String!');
    expect(printSchema(transformedSchema)).toMatchSnapshot();

    const result = await execute({
      schema: transformedSchema,
      document: parse(/* GraphQL */ `
        {
          books {
            title
            author
          }
        }
      `),
    });
    expect(result.data.books).toEqual([
      { title: 'abc', author: 'abra' },
      { title: 'def', author: 'cadabra' },
    ]);
  });

  it('should replace correctly field Type with additional type definitions', async () => {
    const transform = new ReplaceFieldTransform({
      config: {
        typeDefs: /* GraphQL */ `
          type NewAuthor {
            age: String
          }
        `,
        replacements: [
          {
            from: {
              type: 'Query',
              field: 'books',
            },
            to: {
              type: 'BooksApiResponse',
              field: 'books',
            },
            scope: 'hoistValue',
          },
          {
            from: {
              type: 'Author',
              field: 'age',
            },
            to: {
              type: 'NewAuthor',
              field: 'age',
            },
          },
        ],
      },
      cache,
      pubsub,
      baseDir,
      apiName: '',
      syncImportFn: require,
    });
    const schema = makeExecutableSchema({
      typeDefs: schemaDefs,
      resolvers: {
        Query: {
          books: () => ({
            books: [
              { title: 'abc', author: { age: 50 } },
              { title: 'def', author: {} },
            ],
          }),
        },
      },
    });
    const transformedSchema = transform.transformSchema(schema);

    expect((transformedSchema.getType('Author') as GraphQLObjectType).getFields().age.type.toString()).toBe('String');
    expect(printSchema(transformedSchema)).toMatchSnapshot();

    const result = await execute({
      schema: transformedSchema,
      document: parse(/* GraphQL */ `
        {
          books {
            title
            author {
              age
            }
          }
        }
      `),
    });
    expect(result.data.books).toEqual([
      { title: 'abc', author: { age: '50' } },
      { title: 'def', author: { age: null } },
    ]);
  });

  it('should replace correctly field with composer wrapping resolver function', async () => {
    const transform = new ReplaceFieldTransform({
      config: {
        replacements: [
          {
            from: {
              type: 'Query',
              field: 'books',
            },
            to: {
              type: 'BooksApiResponse',
              field: 'books',
            },
            composer: join(__dirname, './fixtures/composer.js#books'),
          },
        ],
      },
      cache,
      pubsub,
      baseDir,
      apiName: '',
      syncImportFn: require,
    });
    const schema = makeExecutableSchema({
      typeDefs: schemaDefs,
      resolvers: {
        Query: {
          books: () => ({
            availableBooks: [{ title: 'abc' }, { title: 'def' }],
          }),
        },
      },
    });
    const transformedSchema = transform.transformSchema(schema);

    const result = await execute({
      schema: transformedSchema,
      document: parse(/* GraphQL */ `
        {
          books {
            title
          }
        }
      `),
    });
    expect(result.data.books).toEqual([{ title: 'abc' }, { title: 'def' }]);
  });
});
