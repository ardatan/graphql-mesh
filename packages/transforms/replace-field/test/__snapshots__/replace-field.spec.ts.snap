// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replace-field applies multiple replaces to obtain a cleaner schema 1`] = `
"type Query {
  books: [Book]
}

type Book {
  title: String!
  author: String!
  isAvailable: Boolean
}
"
`;

exports[`replace-field should replace correctly field Type only 1`] = `
"type Query {
  books: [Book]
}

type Book {
  title: String!
  author: Author!
  code: String
}

type Author {
  name: String!
  age: Int!
}
"
`;

exports[`replace-field should replace correctly field Type with additional type definitions 1`] = `
"type Query {
  books: [Book]
}

type Book {
  title: String!
  author: Author!
  code: String
}

type Author {
  name: String!
  age: String
}
"
`;

exports[`replace-field should replace correctly mtultiple fields with hoistValue and defined resolver function as well as default-field-resolver 1`] = `
"type Query {
  books: [Book]
}

type Book {
  title: String!
  author: String!
  code: String
}
"
`;

exports[`replace-field should replace correctly whole field config 1`] = `
"type Query {
  \\"\\"\\"Retrieve a list of Books\\"\\"\\"
  books(maxResults: Int, orderBy: String): [Book]
}

type Book {
  title: String!
  author: Author!
  code: String
}

type Author {
  name: String!
  age: Int!
}
"
`;
